Map rotation

huss opened the pull request:
# Description

This allows for maps with arbitrary rotations and fixes the circle size issue. In more detail:

- Maps can now be rotated by an arbitrary angle from true north.
- The circle size is scaled by area and made a fraction of the smaller map dimension.
- The calibration error calculation was found to be flawed. It is now a relative error percent.
- Significant documentation was added.

Fixes #571 . Partly Addresses #543 .

## Type of change


- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

- Both the angle and maximum circle size are hard coded. These need to be changed to come from the database/Redux state. The angle is currently set to 0 so it assumes the map is true north. The value is set in calibration.ts and must be changed manually if the map is not true north. That will go away once the DB value is used. The max circle size is set to be 0.15 for now. This value might need to be tweaked and will become the DB default value.
- The code will not work if the map crosses the line on the globe where it goes from -180 to +180 in longitude. This is a discontinuity that messes up the assumed linear nature of the GPS coordinates. It may not be too hard to fix this up but looking at the globe indicates very few parts of the world's land mass cross this point. For now it is not addressed and should be noted in the map documentation.

huss committed:
enhance maps

- Has a lot of the code needed so maps do not need to be oriented north.
The angle is hard codes and needs someone to get in DB and allow admin
to set the value.
- Now sets the size of cicles so they are about a max of 20% of the width
of the graph. It also keeps circles a min size so they do not disappear.
- The code has been tested by not carefully. That is coming but time
to commit this.

huss committed:
inermediate work that may not be correct

huss committed:
intermediate work in progress

huss committed:
map rotation and other map stuff

- The rotation of maps now seems to work fine. This required a lot of
changes to how the map was rotated/shifted and completely changed
how the check occurs for if meter is on the map.
- The circle size calculation was fixed.
- The calibration error is now the relative error. It was not correct before.
- Many comments were added.
- Note the map angle is hard coded and needs to come
from the DB. For now set to 0 so no rotation.
- Note the max circle size is hard coded and should come from admin
at some point.
- Note this will not work if you cross the -180/180 longitude but this
is not as likely. It can happen in a small part of northeast Russia and
Taveui Island and islands on the same longitude that may be nearby.
I am not going to deal with this unlikely case until needed.

huss committed:
fix null gps check

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into mapRotation

limh0228 reviewed (commented):
The code looks good to me. Regarding line 41, I pulled up the code with no map, and it gave a blank screen as expected. 

Perhaps the function gpsToUserGrid can be moved to calibration.ts since it is imported and referenced from multiple files, and is mostly a helper function. (Up to you, you know the code best).

I double checked the coordinate math performed for map rotation starting on line 311 of calibration.ts.

(Very much appreciating the thorough documentation). This code needs a review from someone more acquainted with maps though, as I don't understand all of this code, particularly, the changes to the function meterMapInfoOk, which seem to contain significant differences, and the code calculating scale on line 263 of calibration.ts

huss committed:
move gpsToUserGrid function

- Done per suggestion in PR review.

huss committed:
remove space

huss commented:
I refactored function as suggested by @limh0228. Good idea.

lindavin commented on src/client/app/utils/calibration.ts:
```diff
@@ -86,25 +122,82 @@ export function isValidGPSInput(input: string) {
+ * @param imageDimensions The dimensions of the original map to use from the user.
+ * @returns The error and the origin & opposite point in GPS to use for mapping.
+ */
+export function calibrate(calibrationSet: CalibratedPoint[], imageDimensions: Dimensions): CalibrationResult {
```
I suppose I'm being nitpicky here, but I think there's supposed to be a short summary before the params like for the other functions.

lindavin commented on src/client/app/utils/calibration.ts:
```diff
@@ -200,3 +324,103 @@ export function normalizeImageDimensions(dimensions: Dimensions) {
+	return shifted;
+}
+
+/**
```
For consistency, you might want to include a one line summary of the purpose of this function.

lindavin commented on src/client/app/utils/calibration.ts:
```diff
@@ -200,3 +324,103 @@ export function normalizeImageDimensions(dimensions: Dimensions) {
+	return shiftRotate(size, origin, -1, trueNorthAngle.angle);
+}
+
+/**
```
For consistency, you might want to include a one line summary of the purpose of this function.

huss replied:
Done. Thanks.

lindavin commented on src/client/app/utils/calibration.ts:
```diff
@@ -200,3 +324,103 @@ export function normalizeImageDimensions(dimensions: Dimensions) {
+	return shifted;
+}
+
+/**
```
For consistency, you might want to include a one line summary of the purpose of this function.

huss replied:
Done. Thanks.

lindavin commented on src/client/app/utils/calibration.ts:
```diff
@@ -86,25 +122,82 @@ export function isValidGPSInput(input: string) {
+	return userGrid;
+}
+
+/**
```
I suppose I'm being nitpicky here, but I think there's supposed to be a short summary before the params like for the other functions.

huss replied:
Done. Thanks.

lindavin commented on src/client/app/containers/MapChartContainer.ts:
```diff
@@ -88,18 +158,23 @@ function mapStateToProps(state: State) {
 	const layout: any = {
+		// Either the actual map name or text to say it is not available.
 		title: {
 			text: (map) ? map.name : 'There\'s not an available map'
```
I'm not sure if this text ('There\\'s not an available map') is displayed in the actual UI anywhere, but it might have to be internationalized?

huss replied:
Issue #574 will address this as there are a number of places that maps do not use internationalized strings.

lindavin reviewed (changes requested):
This looks could to me.  I left some minor comments about documentation and another one about internatlizing a string.  Like @limh0228 I am not well aquainted with maps, but I don't see any glaring issues. 
Thank you for the work!

huss committed:
add comments per PR review.

huss commented:
This was reviewed by both @limh0228  and @lindavin but neither was marked as approving. It probably happened since both were doing this. Given no major issues were found and I addressed all the comments, I am going to merge this PR (I cannot approve my own review so doing as admin).
